Add Activity detail screen

SpacyRicochet opened the pull request:
Fixes #26 

This adds an activity detail screen that ticks the following boxes:

* Shows title, start time.
* Shows location, if available.
* Shows mentor, if available.
* Shows description, if available.
* Improved support for Dynamic Type.
* Improved support for Voice Over.

![workshop-detail-screen](https://user-images.githubusercontent.com/904596/66042551-0576e100-e51d-11e9-9325-0cb4e2d686e2.gif)

![IMG_FCE811A25F05-1](https://user-images.githubusercontent.com/904596/66042554-0871d180-e51d-11e9-9b03-1f35f0014a05.jpeg)

SpacyRicochet committed:
Separate Map and Timeline storyboards.

This will help avoid merge conflicts now that more people are working on the project.

SpacyRicochet committed:
Replace schedule title view with navigation title.

SpacyRicochet committed:
Convert header and footer views to cells.

This allows us to better take advantage of UITableView's self-sizing, which doesn't work for header and footer views.

SpacyRicochet committed:
Add workshop detail screen.

SpacyRicochet committed:
Add tag view for area.

SpacyRicochet committed:
Add better VoiceOver support.

SpacyRicochet committed:
Merge branch 'upstream_develop' into 26_timeline-detail

# Conflicts:
#	SwiftIsland.xcodeproj/project.pbxproj
#	SwiftIsland/Timeline/Timeline.storyboard

SpacyRicochet committed:
Merge branch 'upstream_develop' into 26_timeline-detail

SpacyRicochet committed:
Rename to ActivityDetailsViewController.

SpacyRicochet committed:
Adjust lables on dynamic type change.

SpacyRicochet committed:
Adjust area label to better fit dynamic type.

SpacyRicochet committed:
Use better API and make mentor manager injectable.

ppeelen commented on SwiftIsland/Timeline/ActivityDetailsViewController.swift:
```diff
@@ -0,0 +1,121 @@
+    didSet {
+      guard let text = text, text.isEmpty == false else {
+        return
+      }
```
This is just a minor comment, and maybe just a codingstyle taste, but eitherway: In this project, whenever a guard just returns it is written on one line. E.g. `guard let foobar = barfoo else { return }`

SpacyRicochet replied:
Sure. I prefer the latter, so it's easier to breakpoint inside the return. But the single line around is also fine.

ppeelen commented on SwiftIsland/Timeline/TimelineViewController.swift:
```diff
@@ -151,64 +158,94 @@ private extension TimelineViewController {
     if isConcurrent {
-      guard let cell = tableView.dequeueReusableCell(withIdentifier: "ConcurrentCell") as? ConcurrentCell else { return UITableViewCell() }
+      guard let cell = tableView.dequeueReusableCell(withIdentifier: "ConcurrentCell") as? ConcurrentCell else {
+        fatalError("This should always return a cell.")
```
Although this should never happen, Wouldn't it be better to be on the safe side and return `UITableViewCell()`, for the end user I think they might prefer a blank cell over an app crash, right?

SpacyRicochet replied:
I'd argue otherwise; the 'should never happen' indicates that it should never be exposed even to the end user on accident. With a hard crash, you can be pretty sure you'll catch this during development and properly fix it. With a blank cell, you might just think `¯\_(ツ)_/¯` or not even notice it.

I can change this for an `assertFailure(...); return UITableViewCell()`, since that has the same effect, but for actual programming errors, I still prefer `fatalError(...)`.

Let me know.

ppeelen replied:
Sound like a great idea; best of both worlds.

ppeelen reviewed (changes requested):
Nice work! Good page to have and I see so many possibilities on how we can extend on this page.

**Requested change:**
When returning from the activity page, the timeline automatically scrolls down to the current active time. If the user chose to show the details, it should return to the location from where the user left the page, not scroll.

**Discussion:**
Why not use a TableView for `ActivityDetailsViewController`? Now we basically make the same but manually in a UIScrollView. Sure we are not adding a listing/collection of data, but the enduser of it might be just a good with a tableview and it might make it easier to maintain and customise in the future; e.g. if we want to add a map in the header of the tableview, it'll make it easier. What do you think?

SpacyRicochet commented:
Thanks for the review; actual code stuff in their respective comments.

As for **'Why not a UITableViewController?'**, I don't agree with that making it easier to implement or easier to modify/extend. 

For one, making it a table view means I have to start worrying about creating a data source that deals with indexes, for something that doesn't actually have to deal with that.

As for extending, table views often make it harder to do other things and come with unintended side effect. For example, I had to redo the sections of the timeline to allow the header and footer to listen properly to dynamic type, without having to resort to replacing them constantly.

I believe this should be easy enough to manage, especially since it's all contained in the storyboard, but still properly hidden per part of the information.

Hope that explains it sufficiently.

SpacyRicochet committed:
Merge remote-tracking branch 'upstream/develop' into 26_timeline-detail

SpacyRicochet committed:
Conform to repository code guidelines.

Single statement returns or exits should come on the same line as the `else` of a `guard/else` statement.

ppeelen commented:
@spacyRicochet Fair enough, it was worth a discussion and you made some valid arguments which I believe are not worth discussing further!
Thank you for your contribution 🌟🌟🌟

SpacyRicochet committed:
Use slightly safer `assert + return` method here.

SpacyRicochet committed:
Don't scroll to active index after viewing activity detail.
